File: maniaclab.cpp
/**********************************************************************
File name: maniaclab.cpp
This file is part of: ManiacLab

LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>.

FEEDBACK & QUESTIONS

For feedback and questions about ManiacLab please e-mail one of the
authors named in the AUTHORS file.
**********************************************************************/

#include <iostream>
#include <fstream>
#include <cassert>
#include <unistd.h>
#include <cfenv>

#include <boost/python.hpp>

#include "CEngine/IO/Log.hpp"
#include "CEngine/IO/FileStream.hpp"
#include "CEngine/IO/StdIOStream.hpp"
#include "CEngine/IO/Thread.hpp"

#include "CEngine/WindowInterface/Display.hpp"
#include "CEngine/WindowInterface/Window.hpp"
#include "CEngine/WindowInterface/X11/X11Display.hpp"

#include "CEngine/PythonInterface/Package.hpp"
#include "CEngine/PythonInterface/CairoHelpers.hpp"

#include "CManiacLab/PythonInterface.hpp"

using namespace boost::python;
using namespace PyEngine;

PyEngine::Display *disp = 0;

int main(int argc, char** argv) {
    int exitCode = 0;

    // to avoid name conflict with log from cmath
    LogServer *const log = PyEngine::log;

    StreamHandle xmlFile(new FileStream("log.xml", OM_WRITE, WM_OVERWRITE));
    log->addSink(LogStdOutSink(All & (~Debug)));
    log->logf(Debug, "Set up stdout sink");

    log->addSink(LogSinkHandle(
        new LogXMLSink(All & (~Debug), xmlFile, "log.xsl", "ManiacLab")
    ));
    log->logf(Debug, "Set up xml sink");
    log->logf(Information, "Log system started up successfully.");
    try
    {
        log->logf(Information, "Setting up boost/python environment");
        addCUniToInittab();
        addManiacLabToInittab();
        log->logf(Information, "Initializing python");
        Py_Initialize();
        PySys_SetArgv(argc, argv);

        // this must happen after python was initialized. We're loading
        // a module here ;)
        log->logf(Information, "Load some cairo helpers");
        setupCairoHelpers();

        log->logf(Information, "Setting up python runtime");
        object cuni_window_namespace = import("_cuni_window").attr("__dict__");
        object cuni_log_namespace = import("_cuni_log").attr("__dict__");
        object main_namespace = import("__main__").attr("__dict__");

        // FIXME: Is this possible without explicit reference to the
        // platform?
        // Boost needs the explicit type for casting, but it would be
        // nice to force it somehow to do the right thing.
        log->getChannel("platform")->logf(Information, "Setting up X11 environment");
        X11Display *x11 = new X11Display();
        disp = x11;

        cuni_window_namespace["display"] = x11;
        cuni_log_namespace["server"] = logHandle;

        log->logf(Information, "Loading python main module");
        std::string str;
        {
            std::stringstream s;
            std::ifstream main("maniaclab.py");
            if (!main.good()) {
                log->logf(Panic, "Could not find python main file. Terminating.");
                exitCode = 2;
                goto error;
            }
            s << main.rdbuf();
            main.close();
            str = std::string(s.str());
        }
        log->logf(Information, "Handing over control to python");
        exec(str.c_str(), main_namespace);
    }
    catch (boost::python::error_already_set const&)
    {
        PyErr_Print();
        exitCode = 1;
        goto error;
    }
    catch (Exception const& err)
    {
        log->logException(err, Panic);
        exitCode = 1;
        goto noPythonError;
    }

    error:
    // XXX: This is neccessary as python won't free it soon enough for the
    // finalizing writes to happen.
    delete log;
    noPythonError:
    return exitCode;
}

// Local Variables:
// c-file-style: "k&r"
// c-basic-offset: 4
// End:

File: maniaclab.py
#!/usr/bin/python2
# encoding=utf8
# File name: maniaclab.py
# This file is part of: ManiacLab
#
# LICENSE
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or (at
# your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# FEEDBACK & QUESTIONS
#
# For feedback and questions about ManiacLab please e-mail one of the
# authors named in the AUTHORS file.
########################################################################

from __future__ import unicode_literals, print_function, division
from our_future import *

# global PyOpenGL flags MUST ONLY be set here.
import OpenGL
OpenGL.ERROR_ON_COPY = True

if __name__ == '__main__':
    import argparse
    import sys, os
    sys.path.append(os.path.join(os.getcwd(), "PyEngine"))
    import Engine.CEngine.Window as CWindow
    from Engine.CEngine.Log import server as log, Severity

    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-p", "--profile",
        dest="profile",
        nargs="?",
        const=True,
        default=False,
        metavar="FILE",
        help="Run in cProfile mode. If FILE is given, the cProfile\
 output is sent to that file. If FILE is omitted and --profile-shell is\
 set, a default of /tmp/pyuniverse-profile is assumed."
        )

    parser.add_argument(
        "-s", "--profile-shell",
        dest="profileShell",
        action="store_true",
        help="In profiling mode, drop into a shell after the\
 application terminated. In that shell, the profiling data will be\
 available in a pstats variable called p."
        )

    parser.add_argument(
        "--no-threaded-simulation",
        dest="threaded_simulation",
        default=True,
        action="store_false",
        help="Disable multithreading for the simulation."
        )
    args = parser.parse_args(sys.argv[1:])

    from ManiacLab.Application import ManiacLab
    log.log(Severity.Information, "Python is ready to rumble!")

    app = ManiacLab(args)
    if args.profile:
        import cProfile
        log.log(Severity.Warning, "Running in cProfile mode")
        if args.profileShell and args.profile is True:
            log.log(Severity.Warning, "No output file defined but shell requested. Defaulting to /tmp/pyuniverse-profile")
            args.profile = "/tmp/pyuniverse-profile"
        # if args.profileFrames:
        #     if args.profileFrames < 0:
        #         raise ValueError("Nice try.")
        #     app._eventLoop.setFrameCount(args.profileFrames)
        #     log.log(Severity.Information, "Will terminate after {0} frames.".format(args.profileFrames))
        if args.profile is not True:
            log.log(Severity.Information, "cProfile output is going to file: {0}".format(args.profile))
            cProfile.run("app.run()", filename=args.profile)
        else:
            cProfile.run("app.run()")
        if args.profileShell:
            import code
            import pstats
            import readline
            p = pstats.Stats(args.profile)
            def topTimes(n=10):
                p.sort_stats("cum").print_stats(n)
            namespace = {}
            namespace["p"]          = p
            namespace["app"]        = app
            namespace["topTimes"]   = topTimes
            code.InteractiveConsole(namespace).interact("Profiling shell. Stats of the current run are available in the p object. Application state is available in the app object.")
    else:
        app.run()
